### ISSUE TITLE
different validation graph

### ISSUE THREAD

**Issue by MrWanter** [ppwwyyxx/tensorpack#639]
Say I want to modify the shufflenet example to have a different validation graph just like dropout, I observe in the example: https://github.com/ppwwyyxx/tensorpack/blob/e515ad88b71e350e9d9e046786d007e8cf486cff/examples/cifar-convnet.py
`is_training` flag can be got and then used in the `_build_graph`() function:

`is_training = get_current_tower_context().is_training`

so should I just use this and add in `get_logits()` of `shufflenet` example ? do I have to modify the inference callback for validation, as in the `DataParallelInferenceRunner`, the `self.trainer.tower_func` seems to be the training graph
```
    def _setup_graph(self):
        self._handles = []

        assert self.trainer.tower_func is not None, "You must set tower_func of the trainer to use InferenceRunner!"
        input_callbacks = self._input_source.setup(self.trainer.inputs_desc)
        with tf.variable_scope(tf.get_variable_scope(), reuse=True):
            for idx, t in enumerate(self._gpus):
                tower_name = self._tower_names[idx]
                SimplePredictBuilder(
                    ns_name=tower_name,
                    vs_name=self.trainer._main_tower_vs_name, device=t).build(
                        self._input_source, self.trainer.tower_func)
                self._handles.append(self.trainer.tower_func.towers[-1])
```

**Comment by ppwwyyxx** [comment]
In inference it calls `build_graph` with `is_training=False`. So you can use it to build a different model.
`build_graph` is defined in `imagenet_utils.py` which calls `get_logits`.

**Comment by MrWanter** [comment]
But not found where inference called `_build_graph` or `build_graph`, could you pls point out? Thank you

**Comment by ppwwyyxx** [comment]
Depend on what trainer and training interface you use there are many different places.
Starting point: https://github.com/ppwwyyxx/tensorpack/blob/54c5a42db3c33df7310dc80106dd123e4e04d1b4/tensorpack/train/interface.py#L92

**Comment by MrWanter** [comment]
https://github.com/ppwwyyxx/tensorpack/blob/54c5a42db3c33df7310dc80106dd123e4e04d1b4/tensorpack/train/interface.py#L90-L92
but seems this is only for training, while for validation, where does it being called ?

**Comment by ppwwyyxx** [comment]
https://github.com/ppwwyyxx/tensorpack/blob/master/tensorpack/callbacks/inference_runner.py#L135-L140

trainer.tower_func is what you set in trainer.setup_graph()

**Comment by MrWanter** [comment]
I see, but in` _trigger()`:
https://github.com/ppwwyyxx/tensorpack/blob/9bbdf94da8be9976725fb131bfb5c9f5523b25ac/tensorpack/callbacks/inference_runner.py#L261-L278
seems `_set_graph()` or `set_graph` is not called, where am I wrong?

**Comment by ppwwyyxx** [comment]
I don't know what you're talking about.

trainer.tower_func is what you set in trainer.setup_graph(), which you've called in interface.py.
